coap/adapter.go

```diff
@@ -1,61 +1,257 @@
+// Subscriber represents token-NATS subscription map. Token is represented as string in
+// order to simplify access to subscription as well as to simplify unsubscribe feature.
+type Subscriber map[string]*Observer
+
+// AdapterService struct represents CoAP adapter service.
+type AdapterService struct {
 	obsMap map[string][]Observer
-	logger *zap.Logger
+	logger log.Logger
 	pub    mainflux.MessagePublisher
+	subs   map[string]Subscriber
+	nc     *broker.Conn
+	mc     *manager.ManagerClient
 }
 
-// NewCoAPAdapter creates new CoAP adapter struct
-func NewCoAPAdapter(logger *zap.Logger, pub mainflux.MessagePublisher) *CoAPAdapter {
-	ca := &CoAPAdapter{
+// New creates new CoAP adapter service struct.
+func New(logger log.Logger, pub mainflux.MessagePublisher, nc *broker.Conn, mc *manager.ManagerClient) *AdapterService {
+	ca := &AdapterService{
 		logger: logger,
 		pub:    pub,
 		obsMap: make(map[string][]Observer),
+		subs:   make(map[string]Subscriber),
+		nc:     nc,
+		mc:     mc,
 	}
-
 	return ca
 }
 
-// Serve function starts CoAP server
-func (ca *CoAPAdapter) Serve(addr string) error {
-	ca.logger.Info("Starting CoAP server", zap.String("address", addr))
-	return coap.ListenAndServe("udp", addr, ca.COAPServer())
+// SendMessage method sends processes message and pushes it to NATS.
+func (ca *AdapterService) SendMessage(l *net.UDPConn, a *net.UDPAddr, m *coap.Message) *coap.Message {
+	ca.logger.Log("info", fmt.Sprintf("Got message in sendMessage: path=%q: %#v from %v", m.Path(), m, a))
+	var res *coap.Message
+	if len(m.Payload) == 0 && m.IsConfirmable() {
+		res.Code = coap.BadRequest
+		return res
+	}
+
+	// Uri-Query option ID is 15 (0xf).
+	token, err := ca.token(m.Option(0xf))
+	if err != nil {
+		res.Code = coap.Unauthorized
+		return res
+	}
+
+	cid := prefix + mux.Var(m, channel)
+
+	publisher, err := ca.mc.CanAccess(cid, token)
+	if err != nil {
+		res.Code = coap.Unauthorized
+		return res
+	}
+
+	if m.IsConfirmable() {
+		res = &coap.Message{
+			Type:      coap.Acknowledgement,
+			Code:      coap.Content,
+			MessageID: m.MessageID,
+			Token:     m.Token,
+			Payload:   []byte(""),
+		}
+		res.SetOption(coap.ContentFormat, coap.AppJSON)
+	}
+
+	n := mainflux.RawMessage{
+		Channel:   cid,
+		Publisher: publisher,
+		Protocol:  protocol,
+		Payload:   m.Payload,
+	}
+
+	// Publish to channel. Since other adapters are subscribed to
+	// channel with src.coap.* wildcard, everybody will get message.
+	if err := ca.pub.Publish(n); err != nil {
+		if m.IsConfirmable() {
+			res.Code = coap.InternalServerError
+		}
+		return res
+	}
+
+	if m.IsConfirmable() {
+		res.Code = coap.Changed
+	}
+	return res
 }
 
-// BridgeHandler functions is a handler for messages recieved via NATS
-func (ca *CoAPAdapter) BridgeHandler(nm *broker.Msg) {
-	log.Printf("Received a message: %s\n", string(nm.Data))
+func (ca *AdapterService) notify(nm *broker.Msg) {
+	// TODO Authentication has to be added in order to prevent self-notifying.
```

Comment:
It's not the authentication, but the origin check (if I am the publisher, don't return me my message).

Reply:
@mijicd @drasko What do you suggest to do with notificataion messages? In this implementation, it's left to NATS, but, eather way, there has to be some kind of inline structure to keep connections to be able to notify clients later. Previously, same thing was done with no NATS included at all. To keep it this way or to revert to drasko's implementation?